File: coconut2d-x/controllers/CNTouchController.cpp
//
//  CNTouchController.cpp
//  Coconut2D-X
//
//  Created by Moky on 12-11-22.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#include "CNTouchController.h"

#include "../CNNode.h"
#include "../CNScene.h"

NS_CN_BEGIN

USING_NS_CC;

CNTouchController::CNTouchController(void)
: CCObject()
, m_pNode(NULL)
, m_pDelegate(NULL)
, m_tTouchBeganPoint(CCPointZero)
, m_tTouchPoint(CCPointZero)
, m_tTouchEndedPoint(CCPointZero)
, m_tStatus(TouchNodeNormal)
, m_bDragEnabled(false)
, m_tOriginalPosition(CCPointZero)
{
	
}

CNTouchController::~CNTouchController(void)
{
	setDelegate(NULL);
	setNode(NULL);
}

bool CNTouchController::init(void)
{
	setNode(NULL);
	setDelegate(NULL);
	
	setTouchBeganPoint(CCPointZero);
	setTouchPoint(CCPointZero);
	setTouchEndedPoint(CCPointZero);
	
	m_tStatus = TouchNodeNormal;
	
	setDragEnabled(false);
	m_tOriginalPosition = CCPointZero;
	
	return true;
}

CCNode * CNTouchController::getNode(void)
{
	return m_pNode;
}

void CNTouchController::setNode(CCNode * pNode)
{
    if (m_pNode != pNode)
    {
		if (m_pNode)
		{
			CCScene * pScene = CNNode::ancestorScene(m_pNode);
			if (CNScene * scene = dynamic_cast<CNScene *>(pScene))
			{
				scene->removeTouchNode(m_pNode);
			}
		}
        CC_SAFE_RETAIN(pNode);
        CC_SAFE_RELEASE(m_pNode);
        m_pNode = pNode;
		if (m_pNode)
		{
			CCScene * pScene = CNNode::ancestorScene(m_pNode);
			if (CNScene * scene = dynamic_cast<CNScene *>(pScene))
			{
				int iMaxTouches = m_bDragEnabled ? 1 : -1;
				scene->addTouchNode(m_pNode, this, iMaxTouches);
			}
		}
    }
}

bool CNTouchController::ccTouchBegan(CCTouch * pTouch, CCEvent * pEvent)
{
	if (!CNNode::containsTouchLocation(m_pNode, pTouch))
	{
		return false;
	}
	
	// touch points
	CCPoint tPoint = m_pNode->convertTouchToNodeSpace(pTouch);
	setTouchBeganPoint(tPoint);
	setTouchPoint(tPoint);
	setTouchEndedPoint(tPoint);
	
	if (m_bDragEnabled)
	{
		tPoint = m_pNode->convertToWorldSpace(tPoint);
		setTouchBeganPoint(tPoint);
		setTouchPoint(tPoint);
		setTouchEndedPoint(tPoint);
		
		m_tOriginalPosition = m_pNode->getPosition();
		return m_pDelegate ? m_pDelegate->onDragStart(m_pNode) : true;
	}
	
	// button delegate
	m_tStatus = TouchNodeActivie;
	if (m_pDelegate) m_pDelegate->onFingerDown(m_pNode);
	if (m_pDelegate) m_pDelegate->onFingerOver(m_pNode);
	return true;
}

void CNTouchController::ccTouchMoved(CCTouch * pTouch, CCEvent * pEvent)
{
	// touch points
	CCPoint tPoint = m_pNode->convertTouchToNodeSpace(pTouch);
	setTouchPoint(tPoint);
	
	if (m_bDragEnabled)
	{
		tPoint = m_pNode->convertToWorldSpace(tPoint);
		setTouchPoint(tPoint);
		
		CCPoint tDelta = ccpSub(tPoint, m_tTouchBeganPoint);
		
		CCNode * pParent = m_pNode->getParent();
		if (!pParent)
		{
			CNLog("could not happen!");
			return;
		}
		
		CCPoint tOldPosition = m_pNode->getPosition();
		CCPoint tNewPosition = pParent->convertToWorldSpace(m_tOriginalPosition);
		tNewPosition = ccpAdd(tNewPosition, tDelta);
		tNewPosition = pParent->convertToNodeSpace(tNewPosition);
		// try to move to new position
		m_pNode->setPosition(tNewPosition);
		if (m_pDelegate && !m_pDelegate->onDrag(m_pNode))
		{
			m_pNode->setPosition(tOldPosition);
		}
		return;
	}
	
	// button delegate
	if (CNNode::containsTouchLocation(m_pNode, pTouch))
	{
		if (m_tStatus == TouchNodeNormal)
		{
			m_tStatus = TouchNodeActivie;
			if (m_pDelegate) m_pDelegate->onFingerOver(m_pNode);
		}
		else
		{
			if (m_pDelegate) m_pDelegate->onFingerMove(m_pNode);
		}
	}
	else
	{
		if (m_tStatus == TouchNodeActivie)
		{
			m_tStatus = TouchNodeNormal;
			if (m_pDelegate) m_pDelegate->onFingerOut(m_pNode);
		}
	}
}

void CNTouchController::ccTouchEnded(CCTouch * pTouch, CCEvent * pEvent)
{
	// touch points
	CCPoint tPoint = m_pNode->convertTouchToNodeSpace(pTouch);
	setTouchPoint(tPoint);
	setTouchEndedPoint(tPoint);
	
	if (m_bDragEnabled)
	{
		tPoint = m_pNode->convertToWorldSpace(tPoint);
		setTouchPoint(tPoint);
		setTouchEndedPoint(tPoint);
		
		if (m_pDelegate && !m_pDelegate->onDragEnd(m_pNode))
		{
			CCActionInterval * moveTo = CCMoveTo::actionWithDuration(0.2f, m_tOriginalPosition);
			moveTo = CCEaseInOut::actionWithAction(moveTo, 2.0f);
			m_pNode->runAction(moveTo);
		}
		return;
	}
	
	// button delegate
	m_tStatus = TouchNodeNormal;
	if (CNNode::containsTouchLocation(m_pNode, pTouch))
	{
		if (m_pDelegate) m_pDelegate->onFingerUp(m_pNode);
	}
	if (m_pDelegate) m_pDelegate->onFingerOut(m_pNode);
	
	// guesture delegate
	CCPoint d = ccpSub(m_tTouchEndedPoint, m_tTouchBeganPoint);
	CCFloat quantum = 100.0f;
	if (d.x > - quantum && d.x < quantum)
	{
		if (d.y > quantum)
		{
			if (m_pDelegate) m_pDelegate->onSwipeUp(m_pNode);
			return;
		}
		else if (d.y < - quantum)
		{
			if (m_pDelegate) m_pDelegate->onSwipeDown(m_pNode);
			return;
		}
	}
	else if (d.y > - quantum && d.y < quantum)
	{
		if (d.x > quantum)
		{
			if (m_pDelegate) m_pDelegate->onSwipeRight(m_pNode);
			return;
		}
		else if (d.x < - quantum)
		{
			if (m_pDelegate) m_pDelegate->onSwipeLeft(m_pNode);
			return;
		}
	}
	
	// click delegate
	if (CNNode::containsTouchLocation(m_pNode, pTouch))
	{
		if (m_pDelegate) m_pDelegate->onClick(m_pNode);
	}
}

void CNTouchController::ccTouchCancelled(CCTouch * pTouch, CCEvent * pEvent)
{
	ccTouchEnded(pTouch, pEvent);
}

NS_CN_END

File: coconut2d-objc/controllers/CNTouchController.m
//
//  CNTouchController.m
//  Coconut2D
//
//  Created by Moky on 12-11-2.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "CNTouchController.h"

#import "cnMacros.h"
#import "CNNode.h"
#import "CNScene+Touch.h"

@implementation CNTouchController

@synthesize node = node_;
@synthesize delegate = delegate_;

@synthesize touchBeganPoint = touchBeganPoint_;
@synthesize touchPoint = touchPoint_;
@synthesize touchEndedPoint = touchEndedPoint_;

@synthesize dragEnabled = dragEnabled_;

- (id) init
{
	if ((self = [super init]))
	{
		self.node = nil;
		self.delegate = nil;
		
		touchBeganPoint_ = CGPointZero;
		touchPoint_ = CGPointZero;
		touchEndedPoint_ = CGPointZero;
		
		status_ = kTouchNodeNormal;
		
		dragEnabled_ = NO;
		originalPosition_ = CGPointZero;
	}
	return self;
}

- (void) dealloc
{
	[delegate_ release];
	[node_ release];
	[super dealloc];
}


#pragma mark -


- (void) setNode:(CCNode *)node
{
	if (node_ != node)
	{
		if ([node_ isKindOfClass:[CCNode class]])
		{
			CCScene * scene = [node_ scene];
			if ([scene isKindOfClass:[CNScene class]])
			{
				[(CNScene *)scene removeTouchNode:node_];
			}
		}
		
		[node_ release];
		node_ = [node retain];
		
		if ([node_ isKindOfClass:[CCNode class]])
		{
			CCScene * scene = [node_ scene];
			if ([scene isKindOfClass:[CNScene class]])
			{
				int maxTouches = dragEnabled_ ? 1 : -1;
				[(CNScene *)scene addTouchNode:node_ withDelegate:self maxTouches:maxTouches];
			}
		}
	}
}

- (void) doEvent:(SEL)s
{
	if ([delegate_ respondsToSelector:s])
	{
		NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
		[delegate_ performSelector:s withObject:node_];
		[pool release];
	}
}

#pragma mark CCTargetedTouchDelegate

- (BOOL) ccTouchBegan:(UITouch *)touch withEvent:(UIEvent *)event
{
	if (![node_ containsTouchLocation:touch])
	{
		return NO;
	}
	
	// touch points
	CGPoint point = [node_ convertsTouchToNodeSpace:touch];
	touchBeganPoint_ = point;
	touchPoint_ = point;
	touchEndedPoint_ = point;
	
	if (dragEnabled_)
	{
		touchBeganPoint_ = [node_ convertToWorldSpace:touchBeganPoint_];
		touchPoint_ = touchBeganPoint_;
		touchEndedPoint_ = touchBeganPoint_;
		
		originalPosition_ = node_.position;
		
		BOOL flag = YES;
		if ([delegate_ respondsToSelector:@selector(onDragStart:)])
		{
			NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
			flag = [delegate_ onDragStart:node_];
			[pool release];
		}
		return flag;
	} // EOF if (dragEnabled_)
	
	// button delegate
	status_ = kTouchNodeActive;
	[self doEvent:@selector(onFingerDown:)];
	[self doEvent:@selector(onFingerOver:)];
	
	return YES;
}

- (void) ccTouchMoved:(UITouch *)touch withEvent:(UIEvent *)event
{
	// touch points
	CGPoint point = [node_ convertsTouchToNodeSpace:touch];
	touchPoint_ = point;
	
	if (dragEnabled_)
	{
		touchPoint_ = [node_ convertToWorldSpace:touchPoint_];
		
		point = [node_ convertToWorldSpace:point];
		CGPoint delta = ccpSub(point, touchBeganPoint_);
		
		CGPoint oldPosition = node_.position;
		CGPoint newPosition = [node_.parent convertToWorldSpace:originalPosition_];
		newPosition = ccpAdd(newPosition, delta);
		newPosition = [node_.parent convertToNodeSpace:newPosition];
		// try to move to new position
		node_.position = newPosition;
		
		// check
		if ([delegate_ respondsToSelector:@selector(onDrag:)])
		{
			NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
			if (![delegate_ onDrag:node_])
			{
				node_.position = oldPosition;
			}
			[pool release];
		}
		return;
	} // EOF if (dragEnabled_)
	
	if (![node_ containsTouchLocation:touch])
	{
		if (status_ == kTouchNodeActive)
		{
			status_ = kTouchNodeNormal;
			[self doEvent:@selector(onFingerOut:)];
		}
		return;
	}
	
	// button delegate
	if (status_ != kTouchNodeActive)
	{
		status_ = kTouchNodeActive;
		[self doEvent:@selector(onFingerOver:)];
	}
	else
	{
		[self doEvent:@selector(onFingerMove:)];
	}
}

- (void) ccTouchEnded:(UITouch *)touch withEvent:(UIEvent *)event
{
	// touch points
	CGPoint point = [node_ convertsTouchToNodeSpace:touch];
	touchPoint_ = point;
	touchEndedPoint_ = point;
	
	if (dragEnabled_)
	{
		touchPoint_ = [node_ convertToWorldSpace:touchPoint_];
		touchEndedPoint_ = touchPoint_;
		
		if ([delegate_ respondsToSelector:@selector(onDragEnd:)])
		{
			NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
			if (![delegate_ onDragEnd:node_])
			{
				[self ccTouchCancelled:touch withEvent:event];
			}
			[pool release];
		}
		return ;
	} // EOF if (dragEnabled_)
	
	if (![node_ containsTouchLocation:touch])
	{
		return ;
	}
	
	// button delegate
	status_ = kTouchNodeNormal;
	[self doEvent:@selector(onFingerUp:)];
	[self doEvent:@selector(onFingerOut:)];
	
	// guestures delegate
	CGFloat dx = touchEndedPoint_.x - touchBeganPoint_.x;
	CGFloat dy = touchEndedPoint_.y - touchBeganPoint_.y;
	if (dx > - 100 && dx < 100)
	{
		if (dy > 100)
		{
			[self doEvent:@selector(onSwipeUp:)];
			return;
		}
		else if (dy < -100)
		{
			[self doEvent:@selector(onSwipeDown:)];
			return;
		}
	}
	else if (dy > -100 && dy < 100)
	{
		if (dx > 100)
		{
			[self doEvent:@selector(onSwipeRight:)];
			return;
		}
		else if (dx < -100)
		{
			[self doEvent:@selector(onSwipeLeft:)];
			return;
		}
	}
	
	// click delegate
	switch ([touch tapCount])
	{
		case 0:
		case 1:
			[self doEvent:@selector(onClick:)];
			//[self performSelector:@selector(singleTap) withObject:nil afterDelay:0.2f];
			break;
		case 2:
			[self doEvent:@selector(onDoubleClick:)];
			//[NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(singleTap) object:nil];
			//[self performSelector:@selector(doubleTap) withObject:nil afterDelay:0.2f];
			break;
		//case 3:
			//[NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(doubleTap) object:nil];
			//[self performSelector:@selector(tripleTap) withObject:nil afterDelay:0.2f];
			//break;
		//case 4:
			//[NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(tripleTap) object:nil];
			//break;
		default:
			CNLog(@"tapCount=%d", [touch tapCount]);
			break;
	}
}

- (void) ccTouchCancelled:(UITouch *)touch withEvent:(UIEvent *)event
{
	if (dragEnabled_)
	{
		CCActionInterval * moveTo = [CCMoveTo actionWithDuration:0.2f position:originalPosition_];
		moveTo = [CCEaseInOut actionWithAction:moveTo rate:2.0f];
		[node_ runAction:moveTo];
		return ;
	}
	[self doEvent:@selector(onFingerUp:)];
}

//- (void) singleTap
//{
//	CNLog(@"---------------------------- singleTap");
//	[self doEvent:@selector(onClick:)];
//}
//
//- (void) doubleTap
//{
//	CNLog(@"---------------------------- doubleTap");
//	[self doEvent:@selector(onDoubleClick:)];
//}
//
//- (void) tripleTap
//{
//	CNLog(@"---------------------------- tripleTap");
//}

@end
